### PR TITLE
Skeleton Ready for Review

### PR DESCRIPTION
**Pull request by martinmilev**
## The directory structure of the project. Some _boilerplate code_ included.
Alt text

### CODE REVIEW

File: src/main/java/com/clouway/persistent/adapter/jdbc/ConnectionProvider.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+
+/**
```
Comment (freekman) [inline src/main/java/com/clouway/persistent/adapter/jdbc/ConnectionProvider.java]: .

**Comment by freekman** [comment]
i cant see the picture

**Comment by freekman** [comment]
LGTM

File: src/main/java/com/clouway/persistent/adapter/jdbc/ConnectionProvider.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  public ConnectionProvider(String db) {
+    this.db = db;
+  }
+
+  @Override
+  public Connection get() {
+    try {
+      Class.forName("com.mysql.jdbc.Driver");
+      return DriverManager.getConnection("jdbc:mysql://localhost/" + db + "?autoReconnect=true&useSSL=false", "root", "123");
```
Comment (mgenov) [inline src/main/java/com/clouway/persistent/adapter/jdbc/ConnectionProvider.java]: database host should not be hardcoded.

File: src/main/java/com/clouway/persistent/adapter/jdbc/ConnectionProvider.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    } catch (ClassNotFoundException | SQLException e) {
+      e.printStackTrace();
```
Comment (mgenov) [inline src/main/java/com/clouway/persistent/adapter/jdbc/ConnectionProvider.java]: You can throw new IllegalStateException("The MySQL JDBC driver wasn't configured") when ClassNotFoundException is thrown to indicate that some configured was missed. 

The SQLConnection need to be re-thrown when connection cannot be established.

Comment (mgenov) [inline src/main/java/com/clouway/persistent/adapter/jdbc/ConnectionProvider.java]: Returning of null is bad, cause when you use the provided you always have to check for null.

File: src/main/java/com/clouway/persistent/datastore/DataStore.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  public void update(String query, Object... objects) {
+    Connection connection = provider.get();
+    try (PreparedStatement statement = connection.prepareStatement(query)) {
+      fillStatement(statement, objects);
+
+      statement.execute();
+    } catch (SQLException e) {
+      e.printStackTrace();
+    } finally {
+      close(connection);
```
Comment (mgenov) [inline src/main/java/com/clouway/persistent/datastore/DataStore.java]: Connection error cases are not handled.

File: src/main/java/com/clouway/persistent/adapter/jdbc/ConnectionProvider.java (hunk 2 above)
Comment (martinmilev) [inline src/main/java/com/clouway/persistent/adapter/jdbc/ConnectionProvider.java]: fixed

File: src/main/java/com/clouway/persistent/adapter/jdbc/ConnectionProvider.java (hunk 3 above)
Comment (martinmilev) [inline src/main/java/com/clouway/persistent/adapter/jdbc/ConnectionProvider.java]: fixed

File: src/main/java/com/clouway/persistent/datastore/DataStore.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    this.provider = provider;
+  }
+
+  public void update(String query, Object... objects) {
+    Connection connection = provider.get();
+    try (PreparedStatement statement = connection.prepareStatement(query)) {
+      fillStatement(statement, objects);
+      statement.execute();
+    } catch (SQLException e) {
+      throw new IllegalStateException("Connection to the database wasn't established");
```
Comment (freekman) [inline src/main/java/com/clouway/persistent/datastore/DataStore.java]: are you sure this here is needed

File: src/main/java/com/clouway/persistent/datastore/DataStore.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    List<T> list = Lists.newArrayList();
+    Connection connection = provider.get();
+    try (PreparedStatement statement = connection.prepareStatement(query)) {
+      ResultSet resultSet = statement.executeQuery(query);
+      while (resultSet.next()) {
+        T row = rowFetcher.fetchRow(resultSet);
+        list.add(row);
+      }
+    } catch (SQLException e) {
+      throw new IllegalStateException("Connection to the database wasn't established");
```
Comment (freekman) [inline src/main/java/com/clouway/persistent/datastore/DataStore.java]: .

**Comment by freekman** [comment]
LGTM

**Pull request merged**